js/ClockCtrl.js

```diff
@@ -1,348 +1,269 @@
-                    $scope.time = t;
-                    $scope.state = 'paused';
-                    $scope.runTrack.pause();
-                    $scope.pauseTime = t/1000;
-                } else {
-                    $scope.start(pauseStamp);
-                }
-            };
+        $scope.pause = function(pauseStamp) {
+            pauseStamp = pauseStamp||(+new Date());
+            if ($scope.state === 'started') {
+                var startTime = ($scope.pauseTime||$scope.armTime);
+                var t = (startTime*1000) - (pauseStamp - ($scope.startStamp||pauseStamp));
+                $scope.time = t;
+                $scope.state = 'paused';
+                $tracks.pause();
+                $scope.pauseTime = t/1000;
+            } else {
+                $tracks.unpause();                
+                $scope.start(pauseStamp);
+            }
+        };
 
-            $scope.start = function(startStamp,countdown) {
-                if (countdown) {
-                    $scope.arm(countdown);
-                }
-                $scope.startStamp = startStamp||(+(new Date()));
-                $scope.state = 'started';
-                $scope.runTrack.play();
-                $scope.tick();
-                $timeout(function() {
-                    $scope.stopTrack.reset();
-                },500);
-            };
+        $scope.start = function(startStamp,countdown) {
+            if (countdown) {
+                $scope.arm(countdown);
+            }
+            $scope.startStamp = startStamp||(+(new Date()));
+            $scope.state = 'started';
+            $tracks.trigger('start', $scope.time);
+            $scope.tick();
+        };
 
-            $scope.stop = function() {
-                $scope.state = 'stopped';
-                $scope.pauseTime = false;
-                $scope.runTrack.reset();
-            };
+        $scope.stop = function() {
+            $scope.state = 'stopped';
+            $scope.pauseTime = false;
+            $tracks.trigger(['end','stop'], $scope.time);
+        };
 
-            $scope.zero = function() {
-                $scope.time = 0;
-                $scope.stopTrack.play();
+        $scope.toggle = function() {
+            if ($scope.state == 'started') {
                 $scope.stop();
-            };
+            } else {
+                $scope.start();
+            }
+        };
 
-            $scope.toggle = function() {
-                if ($scope.state == 'started') {
+        $scope.mode = function() {
+            $scope.tenths = !$scope.tenths;
+        };
+
+        $scope.tick = function() {
+            if ($scope.state === 'started') {
+                var now = +(new Date());
+                var startTime = ($scope.pauseTime||$scope.armTime);
+                $scope.time = (startTime*1000) - (now - $scope.startStamp);
+                $tracks.trigger(Math.floor($scope.time / 1000), $scope.time);

```

Comment:
This triggers on every tick. I would say that time based triggers would only trigger once

Reply:
Yes. Every tick triggers all events that listen to that tick.
For instance, on second 100, the event '100' will be fired. So that if there's a track waiting for th 100th second, it will be triggered.

Reply:
The problem is that on second 100 it will be fired *every 10 ms* (the interval of tick), so 100 times for the same second. It is not apparent now as tracks don't restart while they are already playing (and tracks on seconds is generally broken)

Reply:
I see. I added an if so it will trigger only when the second changes